pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -393,98 +395,105 @@ func (m *topologyPairsPodSpreadMap) addPod(addedPod, preemptorPod *v1.Pod, node
 		return nil
 	}
 
-	// records which topology key(s) needs to be updated
-	minMatchNeedingUpdate := make(map[string]struct{})
 	podLabelSet := labels.Set(addedPod.Labels)
 	for _, constraint := range constraints {
 		if match, err := PodMatchesSpreadConstraint(podLabelSet, constraint); err != nil {
 			return err
 		} else if !match {
 			continue
 		}
-		pair := topologyPair{
-			key:   constraint.TopologyKey,
-			value: node.Labels[constraint.TopologyKey],
-		}
-		// it means current node is one of the critical paths of topologyKeyToMinPodsMap[TopologyKey]
-		if int32(len(m.topologyPairToPods[pair])) == m.topologyKeyToMinPodsMap[pair.key] {
-			minMatchNeedingUpdate[pair.key] = struct{}{}
-		}
-		m.addTopologyPair(pair, addedPod)
-	}
-	// no need to addTopologyPairWithoutPods b/c if a pair without pods must be present,
-	// it should have already been created earlier in removePod() phase
-
-	// In most cases, min match map doesn't need to be updated.
-	// But it's required to be updated when current node is the ONLY critical path which impacts
-	// the min match. With that said, in this case min match needs to be updated to min match + 1
-	if len(minMatchNeedingUpdate) != 0 {
-		// TODO(Huang-Wei): performance can be optimized.
-		// A possible solution is to record number of critical paths which co-impact the min match.
-		tempMinMatchMap := make(map[string]int32, len(minMatchNeedingUpdate))
-		for key := range minMatchNeedingUpdate {
-			tempMinMatchMap[key] = math.MaxInt32
-		}
-		for pair, podSet := range m.topologyPairToPods {
-			if _, ok := minMatchNeedingUpdate[pair.key]; !ok {
-				continue
-			}
-			if l := int32(len(podSet)); l < tempMinMatchMap[pair.key] {
-				tempMinMatchMap[pair.key] = l
-			}
-		}
-		for key, tempMin := range tempMinMatchMap {
-			if tempMin == m.topologyKeyToMinPodsMap[key]+1 {
-				m.topologyKeyToMinPodsMap[key] = tempMin
+
+		k, v := constraint.TopologyKey, node.Labels[constraint.TopologyKey]
+		pair := topologyPair{key: k, value: v}
+		m.tpPairToMatchNum[pair] += 1
+		curMatch := m.tpPairToMatchNum[pair]
+
+		paths := m.tpKeyToCriticalPaths[k]
+		// As addPod() is called after removePod(), only two cases are possible:
+		// 1) v in "paths[*].topologyValue".
+		// 2) v not in "paths[*].topologyValue" - we bypass this case as "curMatch"
+		//    is definitely greater than "paths[1].matchNum".
+		if v == paths[0].topologyValue {
+			paths[0].matchNum = curMatch
+			// paths[0].matchNum has increased by 1,
```

Comment:
in this case, at least 2 keys had the same count, say (4, 4). After adding, we get (5, 4), so we need to swap to (4, 5). This works if those were the 2 only keys with that count. If we had (4, 4, 4), then after adding, the state should be (4, 4, 5).

Reply:
`[2]critialPath` in preemption case doesn't always hold the **minimum** two paths. We only need to guarantee `path[0]` always holds the minimum matchNum, and that's also our predicate algorithm relies on - compare to the minMatch.

So in your case, it's fine that at some point it becomes [4,5] (one of the original 4 gets lost). At anytime we must keep the minMatch tracked well.